src/java/main/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -292,8 +298,20 @@ protected void connectToLeader(InetSocketAddress addr, String hostname)
                 sock.getInputStream()));
         bufferedOutput = new BufferedOutputStream(sock.getOutputStream());
         leaderOs = BinaryOutputArchive.getArchive(bufferedOutput);
-    }   
-    
+    }
+
+    private void createSocket() throws X509Exception, IOException {
```

Comment:
minor style nit: prefer you return the socket from this method and assign to `sock` in the caller.

Reply:
This is also done. Would you please elaborate a little bit on what are the benefits?

Reply:
I prefer pure functions whenever possible. Since they don't touch global state they are trivially thread-safe, and they are also easier to test. Like I said, it's a minor style thing.